File: mergeSort.js
function mergeSort(arr) {
  if(arr.length == 1) {
    return arr;
  }

  let middle = Math.floor(arr.length/2);
  let left = arr.slice(0, middle);
  let right = arr.slice(middle);

  return merge(
    mergeSort(left),
    mergeSort(right)
  )
}

function merge(left, right){
  let leftIndex = 0;
  let rightIndex = 0;
  let result = [];

  while(leftIndex < left.length && rightIndex < right.length) {
    if(left[leftIndex] < right[rightIndex]){
      result.push(left[leftIndex]);
      leftIndex++;
    } else {
      result.push(right[rightIndex]);
      rightIndex++;
    }
  }

  //Here we are merging the array
  return result.concat(left.slice(leftIndex)).concat(right.slice(rightIndex));
}





let sample = [1,6,4,7,5,34];
console.log(mergeSort(sample));

File: mergeSort.py
def mergeSort(arr):
    if (len(arr) < 2):
        return arr

    middle = len(arr)//2
    left = arr[:middle]
    right = arr[middle:]

    leftSort = mergeSort(left)
    rightSort = mergeSort(right)
    return merge(leftSort, rightSort)

def merge(left,right):
    leftIndex = 0
    rightIndex = 0
    result = []

    while leftIndex < len(left) and rightIndex < len(right):
        if left[leftIndex]< right[rightIndex]:
            result.append(left[leftIndex])
            leftIndex+=1
        else:
            result.append(right[rightIndex])
            rightIndex+=1

    result += left[leftIndex: ]
    result += right[rightIndex: ]

    return result




l = [2,1,4,3,6,5]
print(mergeSort(l))
